### ISSUE TITLE
[4.x]: Couldn't backup db

### ISSUE THREAD

**Issue by Vandemonica** [craftcms/cms#15090]
### What happened?

I'm preparing to upgrade my project into `5.x`, however after I performed update to `4.9.5` the backup always failed whether it was the one on `migrate/all` or when I ran `db/backup` command

```
Backing up the database ... error: The shell command "mysqldump "--defaults-file"="C:\Users\Nircahya\AppData\Local\Temp\cvgbekafgzcc.cnf" "--add-drop-table" "--comments" "--create-options" "--dump-date" "--no-autocommit" "--routines" "--default-character-set"="utf8" "--set-charset" "--triggers" "--no-tablespaces" "--column-statistics"="0" "--no-data" "--result-file"="D:\Workspace\www\myproject\storage\backups\craftcms-myproject--2024-05-28-222447--v4.9.5.sql" "my_project_db" && mysqldump "--defaults-file"="C:\Users\Nircahya\AppData\Local\Temp\cvgbekafgzcc.cnf" "--add-drop-table" "--comments" "--create-options" "--dump-date" "--no-autocommit" "--routines" "--default-character-set"="utf8" "--set-charset" "--triggers" "--no-tablespaces" "--column-statistics"="0" "--no-create-info" "--ignore-table"="my_project_db.cs_assetindexdata" "--ignore-table"="my_project_db.cs_cache" "--ignore-table"="my_project_db.cs_imagetransformindex" "--ignore-table"="my_project_db.cs_resourcepaths" "--ignore-table"="my_project_db.cs_phpsessions" "--ignore-table"="my_project_db.cs_sessions" "my_project_db" >> 'D:\Workspace\www\myproject\storage\backups\myproject--2024-05-28-222447--v4.9.5.sql'" failed with exit code 1.
Aborting.
```

I couldn't find anything interesting in the logs but another thing to take a note is I was able to do backup normally on `4.7.1`, let me know if I need to show something

### Craft CMS version

Craft Solo 4.9.5

### PHP version

8.2.19

### Operating system and version

WINNT 10.0

### Webserver and version
Nginx 1.22.1

### Database type and version

MySQL 8.0.32

### Image driver and version

GD 8.2.19

### Installed plugins and versions

- Admin Bar: 4.3.0
- Asset Rev: 7.0.0
- AsyncQueue: 3.2.1
- Blitz Recommendations: 2.2.1
- CP Field Inspect: 1.4.4
- Disqus: 4.0.0
- Dumper: 5.0.1
- Environment Label: 4.0.3
- FastCGI Cache Bust: 4.0.1
- Feed Me: 5.5.0
- Field Manager: 3.0.8
- Google Cloud Storage: 2.2.0
- Logs: 4.0.1
- Navigation: 2.0.27
- Ray: 2.1.1
- Redactor: 3.0.4
- Retcon: 2.7.5
- SEOMate: 2.2.1
- Sherlock: 4.5.0
- Sprig: 2.8.1
- Super Table: 3.0.14
- Template Select: 3.0.0
- Typed link field: 2.1.5
- Wordsmith: 4.4.0

**Comment by timkelty** [comment]
@Vandemonica are you running this via DDEV, or something else?

**Comment by Vandemonica** [comment]
no, just the normal craft command, ran with windows 10 CMD

this one
```
php craft db/backup
```

**Comment by angrybrad** [comment]
@Vandemonica Sorry... does that mean Craft is installed on your native Windows host OS and you've installed Apache or are using IIS and MySQL on the host machine? i.e. no WSL, Lando, WAMP, etc.

I'm unable to reproduce with DDEV on Windows, so trying to get as close to your setup as possible.

**Comment by Vandemonica** [comment]
Kinda, I used Laragon with Nginx 1.22.1 as webserver

**Comment by mmc501** [comment]
I have just upgraded to 4.9.5 from 4.8.something and I am getting the same issue when I try to run a backup.
I also got the same on a Craft 5 installation.

`2024-06-03 13:38:33 [web.ERROR] [yii\base\Exception] yii\base\Exception: Could not create backup: The shell command "mysqldump "--defaults-file"="C:\windows\TEMP\hoxexocbnxig.cnf" "--add-drop-table" "--comments" "--create-options" "--dump-date" "--no-autocommit" "--routines" "--default-character-set"="utf8" "--set-charset" "--triggers" "--no-tablespaces" "--single-transaction" "--column-statistics"="0" "--no-data" "--result-file"="C:\Websites\Sitename\storage\backups\iakoe--2024-06-03-123833--v4.9.5.sql" "sitename" && mysqldump "--defaults-file"="C:\windows\TEMP\hoxexocbnxig.cnf" "--add-drop-table" "--comments" "--create-options" "--dump-date" "--no-autocommit" "--routines" "--default-character-set"="utf8" "--set-charset" "--triggers" "--no-tablespaces" "--single-transaction" "--column-statistics"="0" "--no-create-info" "--ignore-table"="sitename.craft_assetindexdata" "--ignore-table"="sitename.craft_cache" "--ignore-table"="sitename.craft_imagetransformindex" "--ignore-table"="sitename.craft_resourcepaths" "--ignore-table"="sitename.craft_phpsessions" "--ignore-table"="sitename.craft_sessions" "sitename" >> 'C:\Websites\Sitename\storage\backups\iakoe--2024-06-03-123833--v4.9.5.sql'" failed with exit code 1. in C:\Websites\Sitename\vendor\craftcms\cms\src\controllers\UtilitiesController.php:226`

Also running on windows, but just the default IIS & MySQL.
I think this may be the issue: 

"--defaults-file"="C:\windows\TEMP\hoxexocbnxig.cnf"
I don't see that file existing.

When I remove that from the command and run it in a Command Window it backs up fine.

**Comment by brandonkelly** [comment]
Thanks @mmc501! We were able to reproduce, and have fixed it for the next release.

You can test by changing your `craftcms/cms` requirement in composer.json to `4.x-dev as 4.9.6` and running `composer update`.

**Comment by mmc501** [comment]
Thanks @brandonkelly, tried that and issue is resolved. Thanks

**Comment by brandonkelly** [comment]
Thanks for confirming, @mmc501! Craft 4.9.7 and 5.1.9 are out now with that fix.

**Comment by ipetrov87** [comment]
Hi am currently updating my Craft CMS from 4.15.2 to 5.7.3 and got this issue. I am currently on Window 10 with PHP 8.3, Apache and MariaDB 10.6.21. 

When I run the command from error massage in the Cmder I got this error message
```
Could not open required defaults file: C:\Users\****\AppData\Local\Temp\jccexwtplhdl.cnf
```
I check if this is some permission error with running the command with Cmder as administrator but got the same issue. Then I try to find this file, but it was not there. 

It is similar to @mmc501 case.

**Comment by brandonkelly** [comment]
@ipetrov87 Are you able to test if database backups are working on 4.15.2?

**Comment by ipetrov87** [comment]
Yes. It worked on Craft CMS 4.15.2.  

Initially I was thinking that the problem maybe is the database, because when I double check the requirements for Craft 5 are MariaDB 10.4.6+ and my local database was 10.3.x. So I stash all changes that I made in my code and roll back to Craft CMS 4.15.2. Then download a fresh copy of production database and import it to my local. First think that I made after the import was to make a backup, and it work. Then I update my local MariaDB to 10.6.21. Checkout the code changes to Craft CMS 5 and run ```composer update```. I want to be sure that now the backup will be done successfully before I run ```php craft up```, so I run ```php craft db/backup``` and got the same error.

**Comment by angrybrad** [comment]
@ipetrov87 just tested this with DDEV on Windows 11 using MariaDB 10.6, PHP 8.3, and Craft 4.15.2 and 5.7.4 and it worked as expected .

```
D:\dev\craft4.support>ddev ssh
angry@craft4:/var/www/html$ php craft db/backup
Backing up the database ... done
Backup file: /var/www/html/storage/backups/craft4-support-ddev--2025-04-28-173933--v4.15.2.sql (96.594 KB)
angry@craft4:/var/www/html$
```

> When I run the command from error massage in the Cmder I got this error message
> Could not open required defaults file: C:\Users\****\AppData\Local\Temp\jccexwtplhdl.cnf
> I check if this is some permission error with running the command with Cmder as administrator but got the same issue. Then I try to find this file, but it was not there.

Craft writes out that temporary file to store connection credentials to use when making the backup. If there is an error, it will clean up after itself and remove that file to not leave sensitive information on the system.

For debugging purposes, if you want that file to persist so you can view the content/permissions of it, commenting out this line should do it: https://github.com/craftcms/cms/blob/4.x/src/db/Connection.php#L573

That will live in `vendor/craftcms/cms/src/db/Connection.php`

What dev environment are you using? MAMP? WAMP? DDEV? Everything native on the host machine?

**Comment by ipetrov87** [comment]
After I comment the row that @angrybrad mention, I was able make a new test with existing file, and I got another error
```
mysqldump: Couldn't execute 'SHOW FUNCTION STATUS WHERE Db = '***'': Cannot load from mysql.proc. The table is probably corrupted (1728)
```
With a quick search I found out that I miss to execute the final step of updating my MariaDB from 10.3.x. to 10.6.21 the correct way. I was running the "mysql_upgrade" with a double click in Windows explorer, instead in a terminal as Administrator. When I done this the backup command was able to complete successfully. 

Thank you for the quick response and amazing support.
